render/pixman/renderer.c

```diff
@@ -0,0 +1,234 @@
+		.alpha = color[3] * 0xFFFF,
+	};
+
+	pixman_image_t *fill = pixman_image_create_solid_fill(&colour);
+
+	pixman_image_composite32(PIXMAN_OP_SRC, fill, NULL, buffer->image, 0, 0, 0,
+			0, 0, 0, renderer->viewport_width, renderer->viewport_height);
+
+	pixman_image_unref(fill);
+}
+
+static void pixman_scissor(struct wlr_renderer *wlr_renderer,
+		struct wlr_box *box) {
+	assert(false && "todo pixman_scissor");
+}
+
+static bool pixman_render_subtexture_with_matrix(
+		struct wlr_renderer *wlr_renderer, struct wlr_texture *wlr_texture,
+		const struct wlr_fbox *box, const float matrix[static 9],
+		float alpha) {
+	assert(false && "todo pixman_render_subtexture_with_matrix");
+}
+
+static void pixman_render_quad_with_matrix(struct wlr_renderer *wlr_renderer,
+		const float color[static 4], const float matrix[static 9]) {
+	assert(false && "todo pixman_render_quad_with_matrix");
+}
+
+static void pixman_render_ellipse_with_matrix(struct wlr_renderer *wlr_renderer,
+		const float color[static 4], const float matrix[static 9]) {
+	assert(false && "todo pixman_render_ellipse_with_matrix");
+}
+
+static const enum wl_shm_format *pixman_get_shm_texture_formats(
+		struct wlr_renderer *wlr_renderer, size_t *len) {
+	assert(false && "todo pixman_get_shm_texture_formats");
+}
+
+static struct wlr_texture *pixman_texture_from_pixels(
+		struct wlr_renderer *wlr_renderer, enum wl_shm_format wl_fmt,
+		uint32_t stride, uint32_t width, uint32_t height, const void *data) {
+	assert(false && "todo pixman_texture_from_pixels");
+}
+
+static const struct wlr_drm_format_set *pixman_get_dmabuf_render_formats(
+		struct wlr_renderer *wlr_renderer) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	return &renderer->dmabuf_render_formats;
+}
+
+static bool pixman_bind_buffer(struct wlr_renderer *wlr_renderer,
+		struct wlr_buffer *wlr_buffer) {
+	wlr_log(WLR_INFO, "Pixman bind buffer");
+
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+
+	if (renderer->current_buffer != NULL) {
+		wlr_buffer_unlock(renderer->current_buffer->buffer);
+		renderer->current_buffer = NULL;
+	}
+
+	if (wlr_buffer == NULL) {
+		return true;
+	}
+
+	struct wlr_pixman_buffer *buffer = get_buffer(renderer, wlr_buffer);
+	if (buffer == NULL) {
+		buffer = create_buffer(renderer, wlr_buffer);
+	}
+	if (buffer == NULL) {
+		return false;
+	}
+
+	wlr_buffer_lock(wlr_buffer);
+	renderer->current_buffer = buffer;
+
+	return true;
+}
+
+static const struct wlr_renderer_impl renderer_impl = {
+	.begin = pixman_begin,
+	.clear = pixman_clear,
+	.scissor = pixman_scissor,
+	.render_subtexture_with_matrix = pixman_render_subtexture_with_matrix,
+	.render_quad_with_matrix = pixman_render_quad_with_matrix,
+	.render_ellipse_with_matrix = pixman_render_ellipse_with_matrix,
+	.get_shm_texture_formats = pixman_get_shm_texture_formats,
+	.texture_from_pixels = pixman_texture_from_pixels,
+	.get_dmabuf_render_formats = pixman_get_dmabuf_render_formats,
```

Comment:
Hmm. I see the backends use `get_dmabuf_render_formats` to figure out which formats are supported by the renderer. However the Pixman renderer doesn't deal with DMA-BUFs, so it's a little bit weird to use this function.

Not sure what's the best way to address this yet.

Reply:
Yeah there's a check somewhere in the DRM backend that fails if there's no dmabuf formats.